In [1]:
from sympy import *
import numpy as np

In [2]:
t = Symbol('t')
t0 = Symbol('t0')
r = Symbol('r')
theta = Symbol('theta')
D = Symbol('D')
Sigma = Symbol('Sigma')
Z2max = Symbol('Z2max')

In [3]:
# Here's a height function
h = (t0/t)*exp(-r**2/(4*D*t))

# Testing out this height function
LHS = diff(h,t); print(LHS)
RHS = D*diff(h,r,2) + D/r*diff(h,r,1); print(RHS)
LHSmRHS = simplify(LHS-RHS)
print ('difference =', LHSmRHS)

-t0*exp(-r**2/(4*D*t))/t**2 + r**2*t0*exp(-r**2/(4*D*t))/(4*D*t**3)
t0*(-2 + r**2/(D*t))*exp(-r**2/(4*D*t))/(4*t**2) - t0*exp(-r**2/(4*D*t))/(2*t**2)
difference = 0


### In mathematical notation
$h(r,t)=({t_0 \over t})exp(-{r^2 \over 4Dt})$

Regarding normalization, we need to evaluate 


$\int_{0}^\infty \int_{0}^{2 \pi} {h(r,t)rdrd\theta}=2\pi ({t_0 \over t}) \int_{0}^\infty {exp(-{r^2 \over 4Dt})rdr}=4\pi D t_0 $  

Then if we wanted a normalized function, we could use 

$h^\prime (r,t) = ({1 \over 4\pi Dt_0}) h(r,t)  $

In [4]:
# Gradient of the height function
dh_dr = diff(h,r); print(dh_dr)

-r*t0*exp(-r**2/(4*D*t))/(2*D*t**2)


In [5]:
# Square of the gradient is Z^2
Z2 = dh_dr**2; pprint(Z2)

           2 
         -r  
        ─────
 2   2  2⋅D⋅t
r ⋅t₀ ⋅ℯ     
─────────────
      2  4   
   4⋅D ⋅t    


In [6]:
# Curvature of Z^2 
d2Z2_dr2 = diff(Z2,r,2); print(d2Z2_dr2)

t0**2*(2 - 5*r**2/(D*t) + r**4/(D**2*t**2))*exp(-r**2/(2*D*t))/(4*D**2*t**4)


In [7]:
# Finding the derivative of Z2 with respect to t
dZ2_dt = diff(Z2,t); pprint(dZ2_dt)

             2               2 
           -r              -r  
          ─────           ─────
   2   2  2⋅D⋅t    4   2  2⋅D⋅t
  r ⋅t₀ ⋅ℯ        r ⋅t₀ ⋅ℯ     
- ───────────── + ─────────────
       2  5             3  6   
      D ⋅t           8⋅D ⋅t    


In [8]:
# Evaluate dZ^2/dt at the inflection point
dZ2_dt = dZ2_dt.subs(t,Sigma**2/(2*D))
dZ2_dt_at_inflection = dZ2_dt.subs(r,Sigma)
pprint(dZ2_dt_at_inflection)

     3   2  -1 
-24⋅D ⋅t₀ ⋅ℯ   
───────────────
        8      
       Σ       


In [9]:
# Get the inverse of dZ^2/dt
dt_dZ2_at_inflection = 1/dZ2_dt_at_inflection
pprint(dt_dZ2_at_inflection)

      8  
  -ℯ⋅Σ   
─────────
    3   2
24⋅D ⋅t₀ 


In [10]:
# Getting curvature at inflection
d2Z2_dr2_at_inflection = d2Z2_dr2.subs(r,Sigma)
d2Z2_dr2_at_inflection = d2Z2_dr2_at_inflection.subs(t,Sigma**2/(2*D))
pprint(d2Z2_dr2_at_inflection)

     2   2  -1 
-16⋅D ⋅t₀ ⋅ℯ   
───────────────
        8      
       Σ       


In [11]:
# Now getting the probability density
rho_of_t = Sigma*dt_dZ2_at_inflection/d2Z2_dr2_at_inflection
pprint(rho_of_t)

   17  2  
  Σ  ⋅ℯ   
──────────
     5   4
384⋅D ⋅t₀ 


In [12]:
# Creating Z^2 as an independent variable
Z2 = Symbol('Z2')

In [13]:
# Getting the probability density as a function of Z^2
# by substituting Sigma = ((((2*D/exp(1))*t0/Z2)**.5))**.5
rho_of_Z2 = rho_of_t.subs(Sigma,((((2*D/exp(1))*t0/Z2)**.5))**.5)
rho_of_Z2 = simplify(rho_of_Z2)
pprint(rho_of_Z2)

                           4.25   
                     ⎛D⋅t₀⎞      2
0.000706797018943542⋅⎜────⎟    ⋅ℯ 
                     ⎝ Z₂ ⎠       
──────────────────────────────────
               5   4              
              D ⋅t₀               


In [14]:
# Here, setting up the affine approximation to log(rho)
logrho_of_Z2 = log(rho_of_Z2); #print('log(rho) = ',logrho_of_Z2)
pprint(logrho_of_Z2)
dlogrho_of_Z2_dZ2 = diff(logrho_of_Z2,Z2); #print ('d log(rho)/d Z2 = ', dlogrho_of_Z2_dZ2)
pprint(dlogrho_of_Z2_dZ2)

   ⎛                           4.25   ⎞
   ⎜                     ⎛D⋅t₀⎞      2⎟
   ⎜0.000706797018943542⋅⎜────⎟    ⋅ℯ ⎟
   ⎜                     ⎝ Z₂ ⎠       ⎟
log⎜──────────────────────────────────⎟
   ⎜               5   4              ⎟
   ⎝              D ⋅t₀               ⎠
-4.25 
──────
  Z₂  


In [15]:
# Setting up the expansion about Z2max
Z2max = Symbol('Z2max')
f0 = rho_of_Z2.subs(Z2,Z2max); pprint(f0); print('\n')
f1 = dlogrho_of_Z2_dZ2.subs(Z2,Z2max); pprint(f1); print('\n')

Z2max = 1/(2*D*t0*exp(1))
pprint (Z2max); print('\n')
f0 = rho_of_Z2.subs(Z2,Z2max); pprint(f0); print('\n')
f1 = dlogrho_of_Z2_dZ2.subs(Z2,Z2max); pprint(f1); print('\n')

                            4.25   
                     ⎛ D⋅t₀⎞      2
0.000706797018943542⋅⎜─────⎟    ⋅ℯ 
                     ⎝Z2max⎠       
───────────────────────────────────
                5   4              
               D ⋅t₀               


-4.25 
──────
Z2max 


  -1  
 ℯ    
──────
2⋅D⋅t₀


                          4.25   
                  ⎛ 2   2⎞      2
0.942809041582064⋅⎝D ⋅t₀ ⎠    ⋅ℯ 
─────────────────────────────────
               5   4             
              D ⋅t₀              


-8.5⋅ℯ⋅D⋅t₀




In [16]:
# Expressing the slope in terms of the initial gaussian width
f1_of_Sigma = f1/(2*D*t0)*Sigma**2; 
pprint(f1_of_Sigma)

         2
-4.25⋅ℯ⋅Σ 


In [17]:
# Here's the roughness parameter, sigma, calculated from the affine approximation
sigma = (-1/f1_of_Sigma)**.5
pprint(sigma)

                      0.5
                  ⎛1 ⎞   
0.294210585314206⋅⎜──⎟   
                  ⎜ 2⎟   
                  ⎝Σ ⎠   


### In mathematical notation
$f_1 = -4.25 \ e \ \Sigma^2$  

$\sigma = ({1 \over 4.25 e})^{1/2} {1 \over \Sigma}$

As in the 1d case, the wider the initial bump, the less the roughness, independent of the so-called diffusion coefficient. The difference here is that $\sigma$ is a little smaller than in the 1d case, for the same $\Sigma$. 

In [18]:
# # Here, taking the affine approximation to log(rho)
# D_num = 1
# t0_num = .7

# Z2max_num = 1/(2*D_num*t0_num*np.exp(1)); print ('Z2(max) = ', Z2max_num)

# logrho_of_Z2 = log(rho_of_Z2); print('log(rho) = ',logrho_of_Z2)
# dlogrho_of_Z2_dZ2 = diff(logrho_of_Z2,Z2); print ('d log(rho)/d Z2 = ', dlogrho_of_Z2_dZ2)

# f0 = logrho_of_Z2.subs(Z2,Z2max); 
# print('f0 = ')
# pprint(simplify(f0))

In [19]:
# f1 = dlogrho_of_Z2_dZ2.subs(Z2,Z2max); print('f1 = ',simplify(f1))
# f1_of_t0 = f1.subs(Z2max,1/(2*D*t0*exp(1)))
# print('f1(t0) = ',f1_of_t0)
# sigma = (-1/f1_of_t0)**.5
# print('sigma = ',sigma)
# print('Here\'s the big result')
# print('sigma = ')
# pprint(sigma)

# f0 = f0.subs(Z2max,Z2max_num); print('f0 = ', f0)
# f1 = f1.subs(Z2max,Z2max_num); print('f1 = ', f1)

# sigma_num = (-1/f1)**.5; print ('Inflection sigma = ', sigma_num)
# sigma_num = 1/(6.5*D_num*t0_num*exp(1))**.5;  print ('Inflection sigma = ', sigma_num)